examples/widgets/piano.dart

```diff
@@ -47,47 +59,48 @@ class PianoApp extends StatelessComponent {
     new Key(Colors.purple[500], iLoveYou),
   ];
 
-  PianoApp() {
-    loadSounds();
+  Future connect() {
+    return _loadSounds();
   }
 
-  loadSounds() async {
+  Future _loadSounds() async {
     MediaServiceProxy mediaService = new MediaServiceProxy.unbound();
-    shell.requestService(null, mediaService);
-
-    for (Key key in keys) {
-      MediaPlayerProxy player = new MediaPlayerProxy.unbound();
-      mediaService.ptr.createPlayer(player);
-
-      UrlResponse response = await fetchUrl(key.soundUrl);
-      await player.ptr.prepare(response.body);
-      key.player = player;
+    try {
+      shell.requestService(null, mediaService);
+      List<Future> pending = [];
+      for (Key key in keys) {
+        pending.add(key.load(mediaService));
+      }
+      await Future.wait(pending);
+    } finally {
+      mediaService.close();
     }
-    mediaService.close();
-    // Are we leaking all the player connections?
   }
 
   Widget build(BuildContext context) {
     List<Widget> children = [];
     for (Key key in keys) {
-      children.add(
-        new Flexible(
+      children.add(new Flexible(
           child: new Listener(
-            child: new Container(
-              decoration: new BoxDecoration(backgroundColor: key.color)
-            ),
-            onPointerCancel: (_) => key.up(),
-            onPointerDown: (_) => key.down(),
-            onPointerUp: (_) => key.up()
-          )
-        )
-      );
+              child: new Container(
+                  decoration: new BoxDecoration(backgroundColor: key.color)),
+              onPointerCancel: (_) => key.up(),
+              onPointerDown: (_) => key.down(),
+              onPointerUp: (_) => key.up())));
     }
 
     return new Column(children);
   }
 }
 
-void main() {
-  runApp(new PianoApp());
+main() async {
+  PianoApp app = new PianoApp();
+  // use "await" to make sure the sound files are loaded before we show the ui.
```

Comment:
Maybe call `runApp` here with a loading screen?

```dart
runApp(new Center(child: new Text("Loading...")));
```

You can call `runApp` as many times as you like.